### PR TITLE
[REF] Extract code used to render a pseudoconstant when a table is defined.

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Basic extraction in heavily tested code

This permits us to re-use for settings

Before
----------------------------------------
Code in one long function

After
----------------------------------------
Portion extracted out

Technical Details
----------------------------------------

Comments
----------------------------------------

### CODE REVIEW

File: CRM/Core/PseudoConstant.php [hunk 1]
```diff
@@ -264,103 +264,16 @@ public static function get($daoName, $fieldName, $params = [], $context = NULL)
...
-              $I18nParams['context'] = $fieldSpec['localize_context'];
-            }
-            $i18n = CRM_Core_I18n::singleton();
-            $i18n->localizeArray($output, $I18nParams);
-            // Maintain sort by label
-            if ($order == "ORDER BY %2") {
-              $output = CRM_Utils_Array::asort($output);
-            }
-          }
+          $output = self::renderOptionsFromTablePseudoconstant($pseudoconstant, $params, ($fieldSpec['localize_context'] ?? NULL), $context);
```
Comment (demeritcowboy) [inline CRM/Core/PseudoConstant.php]: I assume the brackets are to tease @colemanw ? He he.

Comment (eileenmcnaughton) [reply CRM/Core/PseudoConstant.php]: Did you  see he asked me to put brackets BACK IN in a PR today!

Comment (demeritcowboy) [reply CRM/Core/PseudoConstant.php]: I did. I admit I would have had to look up the precedence. I didn't, so I still don't know.

**Comment by demeritcowboy** [comment]
The $params array gets changed in the original before being passed to the hook but then because there's no ampersand it doesn't get changed in the refactored function. Seems like it could be a problem?

**Comment by eileenmcnaughton** [comment]
@demeritcowboy I have added the ampersand

**Comment by demeritcowboy** [comment]
Maybe we should just always add an ampersand to function params just in case... :grin:

**Comment by eileenmcnaughton** [comment]
aaarrrrghhhh

File: CRM/Core/PseudoConstant.php (hunk 1 above)
Comment (colemanw) [reply CRM/Core/PseudoConstant.php]: I'll take the brackets over `CRM_Utils_Array` any day!

File: CRM/Core/PseudoConstant.php [hunk 2]
```diff
@@ -1557,4 +1470,109 @@ public static function emailOnHoldOptions() {
...
+    }
+    $query .= ' ' . $order;
+    $dao = CRM_Core_DAO::executeQuery($query, $queryParams);
+    while ($dao->fetch()) {
+      $output[$dao->id] = $dao->label;
+    }
+    // Localize results
+    if (!empty($params['localize']) || $pseudoconstant['table'] === 'civicrm_country' || $pseudoconstant['table'] === 'civicrm_state_province') {
+      $I18nParams = [];
+      if ($localizeContent) {
```
Comment (demeritcowboy) [inline CRM/Core/PseudoConstant.php]: This is a sneaky change but is ok. `localizeContext` vs. `localizeContent`. At least it has the proper 'z'. Yeah I said it. I can't hit people playing hockey anymore so this is as close as it gets to sport now.

Comment (eileenmcnaughton) [reply CRM/Core/PseudoConstant.php]: I flipped it back - your point

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by demeritcowboy** [comment]
* General standards
    * [`r-explain`] __PASS__
    * [`r-user`] __PASS__
    * [`r-doc`] __PASS__
    * [`r-run`] __PASS__
      * Don't see anything obvious. The output seem to have the same values before and after.
      * This 'table' variable is weird in that it only seems to come up for three tables, at least for the screens I visited, and they don't seem to have anything to do with the visited screens, so I'm not sure exactly where to go to see what this would affect.
* Developer standards
    * [`r-tech`] __PASS__
    * [`r-code`] __PASS__
    * [`r-maint`] __N/A__
    * [`r-test`] __PASS__
      * The current failure has been noted elsewhere.

**Comment by mattwire** [comment]
Test fail unrelated (conformance). Merging based on @demeritcowboy review

**Pull request merged**